proofs/alf/cvc5/programs/Utils.smt3

```diff
@@ -11,3 +11,81 @@
 ; It could potentially be improved by having a builtin operator, e.g. alf.compare.
 (define compare_var ((T Type :implicit) (U Type :implicit) (a T) (b U))
   (alf.is_neg (alf.add (alf.hash a) (alf.neg (alf.hash b)))))
+
+; Eliminate singleton list, which takes:
+; (1) a function f
+; (2) the identity element of f
+; (3) the term to process
+(program singleton_elim ((T Type) (S Type) (U Type) (f (-> T U S)) (id S) (x S) (x1 T) (x2 U :list))
+  ((-> T U S) S S) S
+  (
+    ((singleton_elim f id (f x1 x2))  (alf.ite (alf.is_eq x2 id) x1 (f x1 x2)))
+    ((singleton_elim f id x)          x)
+  )
+)
+
+;; =============== for ACI_NORM associative, commutative and idempotent
+
+; Append the children of two applications of an ACI operator, which takes:
+; (1) a function f
+; (2) the identity element of f
+; (3) the first term to append
+; (4) the second term to append
+(program ac_append ((T Type) (S Type) (U Type) (f (-> T U S)) (id S) (x S) (x1 T) (x2 U :list) (y1 T) (y2 U :list))
+  ((-> T U S) S S S) S
+  (
+    ((ac_append f id (f x1 x2) (f y1 y2)) (alf.ite (alf.is_eq x1 y1)
+                                            (ac_append f id (f x1 x2) y2)
+                                          (alf.ite (compare_var x1 y1)
+                                            (alf.cons f x1 (ac_append f id x2 (f y1 y2)))
+                                            (alf.cons f y1 (ac_append f id (f x1 x2) y2)))))
+    ((ac_append f id (f x1 x2) id)        (alf.cons f x1 x2))
+    ((ac_append f id id (f y1 y2))        (alf.cons f y1 y2))
+    ((ac_append f id id id)               id)
```

Comment:
Can this be modified to:
```suggestion
    ((ac_append f id x id)                x)
    ((ac_append f id id y)                y)
```
with `(x S :list)` and `(y S :list)`?

Reply:
Yes I think this works. I'll update.

Reply:
Actually, I take it back. The suggested version does not ensure that `x` and `y` are f-applications, which will lead to incorrect behavior upstream. This version insists that they are.

In the end, this will make possible error cases easy to catch, since e.g. `(ac_append and true (or A B) true)` will fail to evaluate, instead of evaluating to `(or A B)`.